mountpoint-s3/tests/fuse_tests/unlink_test.rs

```diff
@@ -0,0 +1,121 @@
+use std::ffi::OsStr;
+use std::fs;
+use std::fs::File;
+use std::io::{ErrorKind, Read, Seek, SeekFrom};
+
+use fuser::BackgroundSession;
+use mountpoint_s3::S3FilesystemConfig;
+use tempfile::TempDir;
+use test_case::test_case;
+
+use crate::fuse_tests::TestClientBox;
+
+/// Simple test cases, assuming a file isn't open for reading elsewhere.
+fn simple_unlink_tests<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, S3FilesystemConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, Default::default());
+
+    // Add a file directly to the bucket
+    test_client.put_object("dir/hello.txt", b"hello world").unwrap();
+    test_client.put_object("dir/foo.txt", b"bar").unwrap();
+
+    let main_dir = mount_point.path().join("dir");
+    let path = mount_point.path().join("dir/hello.txt");
+    let nonexistent_path = mount_point.path().join("dir/not-here.txt");
+
+    // files should be visible in readdir
+    let dir = fs::read_dir(&main_dir).unwrap();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().and_then(OsStr::to_str).unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        vec!["foo.txt", "hello.txt"]
+    );
+
+    let err = fs::remove_file(nonexistent_path).expect_err("file remove/unlink on non-existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    fs::remove_file(&path).expect("file remove/unlink of existing path should succeed");
+
+    let err = fs::remove_file(path).expect_err("file remove/unlink a second time on existing file should fail");
+    assert_eq!(err.kind(), ErrorKind::NotFound);
+
+    // readdir should now show the file as gone
+    let dir = fs::read_dir(&main_dir).unwrap();
+    let dirs: Vec<_> = dir.map(|f| f.unwrap()).collect();
+    assert_eq!(
+        dirs.iter()
+            .map(|f| f.path().file_name().and_then(OsStr::to_str).unwrap().to_owned())
+            .collect::<Vec<_>>(),
+        vec!["foo.txt"]
+    );
+}
+
+#[cfg(feature = "s3_tests")]
+#[test]
+fn simple_unlink_test_s3() {
+    simple_unlink_tests(crate::fuse_tests::s3_session::new, "simple_unlink_tests");
+}
+
+#[test_case(""; "no prefix")]
```

Comment:
Is there any specific reason we perform the test cases with no prefix and with prefix? I have seen it in multiple other places.

Reply:
I'm not super sure on this one. I think generally we just want to make sure everything works when using and not using a prefix.